rerun_cpp/src/rerun/archetypes/image_ext.cpp

```diff
@@ -1,60 +1,70 @@
-        Image(std::vector<datatypes::TensorDimension> shape, datatypes::TensorBuffer buffer)
-            : Image(datatypes::TensorData(std::move(shape), std::move(buffer))) {}
-
-        /// New depth image from tensor data.
-        ///
-        /// Sets the dimension names to "height",  "width" and "channel" if they are not specified.
-        /// Calls `Error::handle()` if the shape is not rank 2 or 3.
-        explicit Image(rerun::components::TensorData _data);
-        // </CODEGEN_COPY_TO_HEADER>
+    // <CODEGEN_COPY_TO_HEADER>
+
+    /// New Image from height/width/channel and tensor buffer.
+    ///
+    /// Sets the dimension names to "height",  "width" and "channel" if they are not specified.
+    /// Calls `Error::handle()` if the shape is not rank 2 or 3.
+    Image(Collection<datatypes::TensorDimension> shape, datatypes::TensorBuffer buffer)
+        : Image(datatypes::TensorData(std::move(shape), std::move(buffer))) {}
+
+    /// New depth image from tensor data.
+    ///
+    /// Sets the dimension names to "height",  "width" and "channel" if they are not specified.
+    /// Calls `Error::handle()` if the shape is not rank 2 or 3.
+    explicit Image(rerun::components::TensorData _data);
+    // </CODEGEN_COPY_TO_HEADER>
 #endif
 
-        Image::Image(rerun::components::TensorData _data) : data(std::move(_data)) {
-            auto& shape = data.data.shape;
-            if (shape.size() != 2 && shape.size() != 3) {
-                Error(
-                    ErrorCode::InvalidTensorDimension,
-                    "Image shape is expected to be either rank 2 or 3."
-                )
-                    .handle();
-                return;
-            }
-            if (shape.size() == 3 && shape[2].size != 1 && shape[2].size != 3 &&
-                shape[2].size != 4) {
-                Error(
-                    ErrorCode::InvalidTensorDimension,
-                    "Only images with 1, 3 and 4 channels are supported."
-                )
-                    .handle();
-                return;
-            }
+    Image::Image(rerun::components::TensorData _data) : data(std::move(_data)) {
+        auto& shape = data.data.shape;
+        if (shape.size() != 2 && shape.size() != 3) {
+            Error(
+                ErrorCode::InvalidTensorDimension,
+                "Image shape is expected to be either rank 2 or 3."
+            )
+                .handle();
+            return;
+        }
+        if (shape.size() == 3 && shape[2].size != 1 && shape[2].size != 3 && shape[2].size != 4) {
+            Error(
+                ErrorCode::InvalidTensorDimension,
+                "Only images with 1, 3 and 4 channels are supported."
+            )
+                .handle();
+            return;
+        }
+
+        // We want to change the dimension names if they are not specified.
+        // But rerun collections are strictly immutable, so create a new one if necessary.
```

Comment:
Are we sure using `Collection` for `TensorDimension` makes sense? When will it avoid allocations and memcpys, realistically?

Reply:
it's less about avoiding copies & allocs but more about being able to use the adapters for this type as well